apollo-router/src/plugins/telemetry/mod.rs

```diff
@@ -1245,6 +1265,120 @@ impl Telemetry {
+            let field_stat = per_type
+                .entry(node.parent_type.clone())
+                .or_default()
+                .per_field_stat
+                .entry(field_name.clone())
+                .or_insert_with(|| metrics::apollo::studio::SingleFieldStat {
+                    return_type: node.r#type.clone(), // not `Default::default()`’s empty string
+                    errors_count: 0,
+                    latency: Default::default(),
+                    estimated_execution_count: 0.0,
+                    requests_with_errors_count: 0,
+                });
+            let latency = Duration::from_nanos(node.end_time.saturating_sub(node.start_time));
+            field_stat
+                .latency
+                .increment_duration(Some(latency), field_execution_weight);
+            field_stat.estimated_execution_count += field_execution_weight;
+            field_stat.errors_count += node.error.len() as u64;
+            if !node.error.is_empty() {
+                field_stat.requests_with_errors_count += 1;
+            }
+        }
+
+        // For example, `field_level_instrumentation_ratio == 0.03` means we send a
+        // `apollo-federation-include-trace: ftv1` header with 3% of subgraph requests.
+        // To compensate, assume that each trace we recieve is representative of 33.3… requests.
+        // Metrics that recieve this treatment are kept as floating point values in memory,
+        // and converted to integers after aggregating values for a number of requests.
+        let field_execution_weight = 1.0 / field_level_instrumentation_ratio;
+
+        let mut per_type = HashMap::new();
+        for trace in traces {
+            if let Some(node) = &trace.root {
+                recur(&mut per_type, field_execution_weight, node)
+            }
+        }
+        per_type
+    }
+
+    fn per_path_error_stats(traces: &[proto::reports::Trace]) -> SinglePathErrorStats {
+        fn recur<'node>(
+            stats_root: &mut SinglePathErrorStats,
+            path: &mut Vec<&'node String>,
+            node: &'node proto::reports::trace::Node,
+        ) {
+            if let Some(ResponseName(name)) = &node.id {
+                path.push(name)
+            }
+            if !node.error.is_empty() {
+                let mut stats = &mut *stats_root;
+                for &name in &*path {
+                    stats = stats.children.entry(name.clone()).or_default();
+                }
+                stats.errors_count += node.error.len() as u64;
+                stats.requests_with_errors_count += 1;
+            }
+            for child in &node.child {
+                recur(stats_root, path, child)
+            }
+            if let Some(ResponseName(_)) = &node.id {
+                path.pop();
+            }
+        }
+        let mut root = Default::default();
+        for trace in traces {
+            if let Some(node) = &trace.root {
+                recur(&mut root, &mut Vec::new(), node)
```

Comment:
This is pretty subtle and stupid but if you look carefully at iterateOverQueryPlan in apollo-server you'll see that for gateway (federated) traces, the path that's produced for these error things starts with `service:SUBGRAPHNAME`.

I think you don't have any snapshot tests that show error stats or that show field execution weights btw.